File: LinkedList/Java/DoublyLL.java
class node
{
  public int data;
  public node next;
  public node prev;

  public node()
  {
    data = 0;
    next = null;
    prev = null;
  }
  public node(int no)
  {
    data = no;
    next = null;
    prev = null;
  }
}

class DLL
{
  private int iCount ;
  private node first;

    public DLL() 
    {
      iCount = 0;
      first = null;
    }
    

    public void Display()
    {
      node temp = first;
      while(temp != null)
      {
        System.out.print("| "+temp.data+" | <=> ");
        temp = temp.next;
      }
      System.out.println(" null");
    }

    public int Count()
    {
      return iCount;
    }
    public void InsertFirst(int no)
    {
      node newn = new  node(no);
      if(first == null)
      {
        first = newn;
      }
      else
      {
        newn.next = first;
        first = newn;

      }
      iCount++;
    }

    public void InsertLast(int no)
    {
      node newn = new node(no);
      node temp = null;
      if(first == null)
      {
        first = newn;
      }
      else
      {
        temp = first;
        while(temp.next != null)
        {
          temp = temp.next;
        }
        temp.next = newn;
        newn.prev = temp;
      }
      iCount++;
    }

    public void InsertAtPos(int no,int pos)
    {
      node newn = new node(no);
      node temp = null;
      if(pos < 1 || pos > iCount+1)
      {
        System.out.println("Invalid position");
        return;
      }

      if(pos == 1)
      {
        InsertFirst(no);
      }
      else if (pos == iCount+1)
      {
        InsertLast(no);
      }
      else 
      {
        temp = first;
      for(int iCnt = 1;iCnt < pos -1;iCnt++)
      {
        temp = temp.next;
      }
      newn.next=temp.next;
      newn.prev = temp;
      temp.next = newn;
      temp.next.prev = newn;
      }
    }

    public void DeleteFirst()
    {
      if(first == null)
      {
        return;
      }
      else
      {
        first = first.next;
        first.prev = null;
      }
    }

    public void DeleteLast()
    {
      node temp = null;
      if(first == null)
      {
        return;
      }
      else
      {
        temp = first;
        while(temp.next.next != null)
        {
          temp = temp.next;
        }
        temp.next = temp.next.next;
        
      }
    }

    public void DeleteAtPos(int pos)
    {
      node temp = null;
      if(pos < 1 || pos > iCount)
      {
        System.out.println("Invalid Position");
        return;
      }

      if(pos == 1)
      {
        DeleteFirst();
      }
      else if ( pos == iCount)
      {
        DeleteLast();
      }
      else
      {
        temp = first;
        for(int iCnt = 1; iCnt < pos -1 ;iCnt ++)
        {
          temp = temp.next;
        }
        temp.next = temp.next.next;
      }
    }
  
}

public class DoublyLL {
  public static void main(String A[]) {
      DLL dobj = new DLL();
      dobj.InsertFirst(21);
      dobj.Display();
      dobj.InsertLast(221);
      dobj.InsertFirst(21);
      dobj.Display();
      dobj.InsertLast(221);
      dobj.Display();
      dobj.InsertAtPos(334, 2);
      // dobj.Display();
      // dobj.DeleteFirst();
      // dobj.Display();
      // dobj.DeleteLast();
      dobj.DeleteAtPos(2);
      dobj.Display();
  }
}

File: LinkedList/Cpp/DoublyLL.cpp
#include<iostream>
using namespace std;

class node 
{
  public:
  int data;
  node * next;
  node * prev ;

  node ()
  {
    data = 0;
    next = nullptr;
    prev = nullptr;
  }

  node (int no)
  {
    data = no;
    next = nullptr;
    prev = nullptr;
  }
};

typedef node NODE;
typedef node * PNODE;



class DoublyLL
{
  private:
  PNODE first;
  int iCount ;

  public:

  DoublyLL()
  {
    first = nullptr;
    iCount = 0;
  }
  void Display()
  {
    PNODE temp = first;
    cout<<"NULL <=> ";
    while(temp != nullptr)
    {
      cout<<"| "<<temp->data<<" | <=> ";
      temp = temp->next;
    }
    cout<<"NULL\n"<<endl;
  }
  void InsertFirst(int no)
  {
    PNODE newn = new NODE(no);
    if(first == nullptr)
    {
      first = newn; 
    }
    else
    {
      newn -> next = first;
      first ->prev = newn;
      first = newn;
    }
    iCount++;
  }


  void InsertLast(int no)
  {
    PNODE newn = new NODE(no);
    PNODE temp = nullptr;
    if(first == nullptr)
    {
      first =  newn;
    }
    else 
    {
      temp = first;
      while (temp -> next != nullptr)
      {
        temp = temp -> next;
      }
      temp -> next = newn ;
      newn -> prev = temp ;
      
    }
    iCount++;
    
  }


    void InsertAtPos(int no,int pos)
    {
      PNODE newn = new NODE(no);
      PNODE temp = nullptr;
      int iCnt = 0;
      if((pos < 1) || (pos > iCount +1))
      {
        cout<<"Invlaid Inpout Position"<<endl;
        return;
      }

      if(pos == 1)
      {
        InsertFirst(no);
      }
      else if ( pos == iCount+1)
      {
        InsertLast(no);
      }
      else
      {
        temp = first ;
        for(iCnt = 1; iCnt < pos -1 ;iCnt++)
        {
          temp = temp -> next;
        }
        newn -> next = temp -> next ;
        temp ->next ->prev = newn;
        newn ->prev = temp;
        temp ->next = newn;

        iCount++;
      }
    }
    


    void DeleteFirst()
    {
      
      if(first == nullptr)
      {
        return;
      }
      else if (first -> next == nullptr)
      {
        delete first;
        first = nullptr;
        
      }
      else
      {
        first = first -> next;
        delete (first -> prev);
        first -> prev = nullptr;
      }
      iCount--;
    }
    void DeleteLast()
    {
      PNODE temp = nullptr;
      if(first == nullptr)
      {
        return;
      }
      else if ( first -> next == nullptr)
      {
        delete first;
        first = nullptr;
      }
      else 
      {
        temp = first;
        while(temp ->next -> next != nullptr)
        {
          temp = temp -> next;
        }
        
        delete temp ->next;
        temp ->next = nullptr;
      }
      iCount--;
    }


    void DeleteAtPos(int pos)
    {
      PNODE temp = nullptr;
      int iCnt = 0;
      if((pos < 1) || (pos > iCount))
      {
        cout<<"Invalid Input Position"<<endl;
        return;
      }
      if(pos == 1)
      {
        DeleteFirst();
      }
      else if ( pos == iCount)
      {
        DeleteLast();
      }
      else
      {
        temp = first;
        for(iCnt = 1 ; iCnt < pos -1 ; iCnt ++)
        {
          temp = temp -> next;
        }
        temp ->next = temp -> next -> next;
        delete temp ->next ->prev;
        temp ->next ->prev = temp;
        iCount--;
      }

    }
    int Count()
    {
      return iCount;
    }

};


int main()
{
  DoublyLL dobj ;
  int iRet = 0;
  dobj.InsertFirst(51);
  dobj.InsertFirst(21);
  dobj.InsertFirst(11);
  dobj.Display();
  iRet = dobj.Count();
  cout<<"number of elements are : "<<iRet<<endl;
  dobj.InsertLast(51);
  dobj.InsertLast(21);
  dobj.InsertLast(11);
  dobj.Display();
  iRet = dobj.Count();
  cout<<"number of elements are : "<<iRet<<endl;
  // dobj.DeleteFirst();
  // dobj.Display();
  // iRet = dobj.Count();
  // cout<<"number of elements are : "<<iRet<<endl;
  // dobj.DeleteLast();
  // dobj.Display();
  // iRet = dobj.Count();
  // cout<<"number of elements are : "<<iRet<<endl;
  dobj.DeleteAtPos(5);
  dobj.Display();
  iRet = dobj.Count();
  cout<<"number of elements are : "<<iRet<<endl;
  dobj.InsertAtPos(105,5);
  dobj.Display();
  iRet = dobj.Count();
  cout<<"number of elements are : "<<iRet<<endl;
  return 0;
}




